specs/DHT.md

```diff
@@ -0,0 +1,399 @@
+# libp2p Daemon Control Protocol
+
+The libp2p daemon DHT protocol allows clients to query and announce to the
+libp2p DHT.
+
+_At the moment, this is a living document. As such, it will be susceptible to
+changes until stabilization._
+
+## Protocol Specification
+
+### Data Types
+
+The data structures are defined in [pb/p2pd.proto](pb/p2pd.proto). All messages
+are varint-delimited. For the DHT queries, the relevant data types are:
+
+- `DHTRequest`
+- `DHTResponse`
+
+All DHT requests will be wrapped in a `Request` message with `Type: DHT`. All
+DHT responses from the daemon will be wrapped in a `Response` with the
+`DHTResponse` field populated.
+
+`DHTRequest` messages have a `Type` parameter that specifies the specific query
+the client wishes to execute.
+
+`DHTResponse` messages have a `Type` parameter that specifies whether a response
+marks the `BEGIN`ning of a stream of messages, a `VALUE` within a stream of
+messages, or the `END` of a stream of messages.
```

Comment:
We should probably expand on this and make specific mention of streams.
If there is a single value returned, then its marked as a `VALUE`.
If it is a stream of values, then the initial response contains `BEGIN`, followed by a stream of `VALUE`s, and finally an `END` message to indicate end of stream.

Reply:
absolutely